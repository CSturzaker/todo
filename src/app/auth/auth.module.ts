import { NgModule } from '@angular/core'
import { CommonModule } from '@angular/common'
import { ReactiveFormsModule } from '@angular/forms'
import { StoreModule } from '@ngrx/store'
import { EffectsModule } from '@ngrx/effects'

// components
import { LoginFormComponent } from './components/login-form/login-form.component'
import { LoginPageComponent } from './containers/login-page/login-page.component'
// effects
import { AuthEffects } from './effects'

// modules
import { AuthRoutingModule } from './auth-routing.module'
import { MaterialModule } from '../material/material.module'
// reducers
import * as fromAuth from './reducers'

@NgModule({
  declarations: [LoginFormComponent, LoginPageComponent],
  imports: [
    CommonModule,
    AuthRoutingModule,
    MaterialModule,
    ReactiveFormsModule,
    StoreModule.forFeature(fromAuth.authFeatureKey, fromAuth.reducers),
    EffectsModule.forFeature([AuthEffects]),
  ],
})
export class AuthModule {}
