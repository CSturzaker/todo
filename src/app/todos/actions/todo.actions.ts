import { v4 as uuid } from 'uuid'
import { createAction, props } from '@ngrx/store'

export const addTodo = createAction(
  '[Todos] Add',
  (name: string, id = uuid()) => ({ name, id })
)

export const toggleTodo = createAction(
  '[Todos] Toggle',
  props<{ id: string }>()
)

export const removeTodo = createAction(
  '[Todos] Remove',
  props<{ id: string }>()
)
