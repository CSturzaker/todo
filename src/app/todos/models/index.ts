export * from './list-item'
export * from './todo'
