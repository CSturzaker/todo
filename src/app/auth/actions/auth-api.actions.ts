import { createAction, props } from '@ngrx/store'

// models
import { User } from '../models'

export const loginSuccess = createAction(
  '[Auth/API] Login Success',
  props<{ user: User }>()
)

export const loginFailure = createAction(
  '[Auth/API] Login Failure',
  props<{ error: any }>()
)

export const loginRedirect = createAction('[Auth/API] Login Redirect')
