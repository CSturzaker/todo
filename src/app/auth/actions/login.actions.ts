import { createAction, props } from '@ngrx/store'

// models
import { Credentials } from '../models'

export const login = createAction(
  '[Login Page] Login',
  props<{ credentials: Credentials }>()
)
