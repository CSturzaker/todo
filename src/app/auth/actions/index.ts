import * as AuthApiActions from './auth-api.actions'
import * as LoginActions from './login.actions'

export { AuthApiActions, LoginActions }
