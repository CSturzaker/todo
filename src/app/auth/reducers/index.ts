import {
  createSelector,
  createFeatureSelector,
  Action,
  combineReducers,
} from '@ngrx/store'
import * as fromRoot from '../../state'
import * as fromAuth from '../reducers/auth.reducer'
import * as fromLoginPage from '../reducers/login-page.reducer'

export const authFeatureKey = 'auth'

export interface AuthState {
  [fromAuth.authFeatureKey]: fromAuth.State
  [fromLoginPage.loginPageFeatureKey]: fromLoginPage.State
}

export interface State extends fromRoot.State {
  [authFeatureKey]: AuthState
}

export function reducers(state: AuthState | undefined, action: Action) {
  return combineReducers({
    [fromAuth.authFeatureKey]: fromAuth.reducer,
    [fromLoginPage.loginPageFeatureKey]: fromLoginPage.reducer,
  })(state, action)
}

export const selectAuthState = createFeatureSelector<State, AuthState>(
  authFeatureKey
)

export const selectAuthStatusState = createSelector(
  selectAuthState,
  (state: AuthState) => state.auth
)
export const selectUser = createSelector(
  selectAuthStatusState,
  fromAuth.getUser
)
export const selectLoggedIn = createSelector(selectUser, user => !!user)

export const selectLoginPageState = createSelector(
  selectAuthState,
  (state: AuthState) => state.loginPage
)
export const selectLoginPageError = createSelector(
  selectLoginPageState,
  fromLoginPage.getError
)
export const selectLoginPagePending = createSelector(
  selectLoginPageState,
  fromLoginPage.getPending
)
export const selectUserName = createSelector(
  selectAuthStatusState,
  fromAuth.getUsername
)
