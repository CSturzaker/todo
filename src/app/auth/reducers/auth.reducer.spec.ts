import { reducer } from './auth.reducer'
import * as fromAuth from './auth.reducer'
import { AuthApiActions } from '../actions'

import { User } from '../models'

describe('AuthReducer', () => {
  describe('undefined action', () => {
    it('should return the default state', () => {
      const action = {} as any

      const result = reducer(undefined, action)

      expect(result).toMatchSnapshot()
    })
  })

  describe('LOGIN_SUCCESS', () => {
    it('should add a user set loggedIn to true in auth state', () => {
      const user = { name: 'test' } as User
      const createAction = AuthApiActions.loginSuccess({ user })

      const result = reducer(fromAuth.initialState, createAction)

      expect(result).toMatchSnapshot()
    })
  })
})
