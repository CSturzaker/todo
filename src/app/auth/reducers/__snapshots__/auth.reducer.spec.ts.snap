// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthReducer LOGIN_SUCCESS should add a user set loggedIn to true in auth state 1`] = `
Object {
  "user": Object {
    "name": "test",
  },
}
`;

exports[`AuthReducer undefined action should return the default state 1`] = `
Object {
  "user": null,
}
`;
